File: python/continuous_subarray_sum.py
class Solution:
    def checkSubarraySum(self, nums: list[int], k: int) -> bool:
        """time complexity: O(n)"""
        remain_map: dict[int, int] = {0: -1}
        total: int = 0
        for i, n in enumerate(nums):
            total += n
            remain: int = total % k
            if remain not in remain_map:
                remain_map[remain] = i
            elif i - remain_map[remain] > 1:
                return True
        return False

    def checkSubarraySum2(self, nums: list[int], k: int) -> bool:
        prefix_sum: int = 0
        prefix_map: dict[int, int] = {0: -1}
        for i in range(len(nums)):
            prefix_sum += nums[i]
            if prefix_sum % k in prefix_map and i - prefix_map[prefix_sum % k] >= 2:
                return True
            if prefix_sum % k not in prefix_map:
                prefix_map[prefix_sum % k] = i
        return False


def test_checkSubarraySum_case_1():
    # arrange
    nums: list = [23, 2, 4, 6, 7]
    k: int = 6
    expected: bool = True

    # act
    actual = Solution().checkSubarraySum(nums, k)

    # assert
    assert actual == expected


def test_checkSubarraySum_case_2():
    # arrange
    nums: list = [23, 2, 6, 4, 7]
    k: int = 13
    expected: bool = False

    # act
    actual = Solution().checkSubarraySum(nums, k)

    # assert
    assert actual == expected


def test_checkSubarraySum_case_3():
    # arrange
    nums: list = [23, 2, 6, 4, 7]
    k: int = 13
    expected: bool = False

    # act
    actual = Solution().checkSubarraySum2(nums, k)

    # assert
    assert actual == expected

File: rust/src/code/continuous_subarray_sum.rs
struct Solution;

impl Solution {
    // time complexity: O(n)
    pub fn check_subarray_sum(nums: Vec<i32>, k: i32) -> bool {
        use std::collections::HashMap;

        let mut remain_map: HashMap<i32, i32> = HashMap::from([(0, -1)]);
        let mut total: i32 = 0;
        for i in 0..nums.len() {
            total += nums[i];
            let remain: i32 = total % k;
            if !remain_map.contains_key(&remain) {
                remain_map.insert(remain, i as i32);
            } else if i as i32 - remain_map[&remain] > 1 {
                return true;
            }
        }
        false
    }
}

#[cfg(test)]
mod test {
    use super::Solution;

    #[test]
    fn test_check_subarray_sum_case_1() {
        // arrange
        let nums: Vec<i32> = vec![23, 2, 4, 6, 7];
        let k: i32 = 6;
        let expected: bool = true;

        // act
        let actual = Solution::check_subarray_sum(nums, k);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_check_subarray_sum_case_2() {
        // arrange
        let nums: Vec<i32> = vec![23, 2, 6, 4, 7];
        let k: i32 = 13;
        let expected: bool = false;

        // act
        let actual = Solution::check_subarray_sum(nums, k);

        // assert
        assert_eq!(actual, expected);
    }
}
